HelloWorld_ST_Sensors/user/gps/gps.cpp

```diff
@@ -0,0 +1,152 @@
+                        break;
+                    case GPGGA_LONGITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.long_dir, temp_token);
+                        break;
+                    }
+                    break;
+                case GPS_GPRMC:
+                    switch (nmea_string_token_index)
+                    {
+                    case GPRMC_VALIDITY_INDEX:
+                        strcpy(gps_coordinates_data.validity, temp_token);
+                        // Consider NMEA strings that are GPRMC and Active Navigation Receiver Warning
+                        if (strcmp(temp_token, "A") != 0)
+                        {
+                            gps_nmea_type = GPS_INVALID_NMEA;
+                        }
+                        break;
+                    case GPRMC_LATITUDE_INDEX:
+                        strcpy(gps_coordinates_data.latitude, temp_token);
+                        break;
+                    case GPRMC_LATITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.lat_dir, temp_token);
+                        break;
+                    case GPRMC_LONGITUDE_INDEX:
+                        strcpy(gps_coordinates_data.longitude, temp_token);
+                        break;
+                    case GPRMC_LONGITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.long_dir, temp_token);
+                        break;
+                    }
+                    break;
+                }
+            }
+        }
+        return gps_nmea_type;
+    }
+
+    uint8_t gps_c::extract_gpgga_or_gprmc_line(char *gps_buff)
+    {
+        uint8_t gps_nmea_type = GPS_INVALID_NMEA;
+        char valid_gps_string[100] = {0};
+        uint16_t counter = 0;
+        for (uint16_t i = 0; i < UART_BUF_LEN; i++)
+        {
+            if (gps_buff[i] == '$')
+            {
+                while (gps_buff[i] != '\n' && i < UART_BUF_LEN)
+                {
+                    valid_gps_string[counter] = gps_buff[i];
+                    i++;
+                    counter++;
+                }
+                if (i < UART_BUF_LEN)
+                {
+                    gps_nmea_type = check_gprmc_gpgga_line(valid_gps_string);
+                    if (gps_nmea_type != GPS_INVALID_NMEA)
+                    {
+                        break;
+                    }
+                }
+                counter = 0;
+                memset(valid_gps_string, 0, strlen(valid_gps_string));
+            }
+        }
+        return gps_nmea_type;
+    }
+
+    void gps_c::register_func(void (*external_fn)(const char *newline))
+    {
+        gps_process_fn_ptr = external_fn;
+    }
+
+    void gps_c::gps_interrupt_cb()
+    {
+        int data = gps_uart.getc();
+        curr_line_buffer[index] = data;
+        if (index != UART_BUF_LEN)
+        {
+            index++;
+        }
+        if (index == UART_BUF_LEN)
+        {
```

Comment:
Why wait for it to be populated with UART_BUF_LEN?
Check for \n and then parse into latitude/longitude or call the registered_callback function

Reply:
The second issue with this logic is that below you are taking the entire buffer of 400 characters and checking for only one received type.
Instead get GPS data per line, the above parsing logic would have much fewer if else lines